src/main/java/com/github/legoatoom/connectiblechains/chain/ChainLink.java

```diff
@@ -0,0 +1,323 @@
+
+import com.github.legoatoom.connectiblechains.ConnectibleChains;
+import com.github.legoatoom.connectiblechains.enitity.ChainCollisionEntity;
+import com.github.legoatoom.connectiblechains.enitity.ChainKnotEntity;
+import com.github.legoatoom.connectiblechains.enitity.ModEntityTypes;
+import com.github.legoatoom.connectiblechains.util.Helper;
+import com.github.legoatoom.connectiblechains.util.NetworkingPackets;
+import io.netty.buffer.Unpooled;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import net.fabricmc.fabric.api.networking.v1.PlayerLookup;
+import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.ItemEntity;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.util.math.Vec3f;
+import net.minecraft.world.GameRules;
+import net.minecraft.world.World;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A logical representation of the link between a knot and another entity.
+ * It also serves as a single source of truth which prevents state mismatches in the code.
+ *
+ * @author Qendolin
+ */
+public class ChainLink {
+    /**
+     * The x/z distance between {@link ChainCollisionEntity ChainCollisionEntities}.
+     * A value of 1 means they are "shoulder to shoulder"
+     */
+    private static final float COLLIDER_SPACING = 1.5f;
+
+    /**
+     * The de facto owner of this link. It is responsive for managing the link and keeping track of it across saves.
+     */
+    @NotNull
+    public final ChainKnotEntity primary;
+    /**
+     * The de facto target of this link. Mostly used to calculate positions.
+     */
+    @NotNull
+    public final Entity secondary;
+    /**
+     * The type of the link
+     */
+    @NotNull
+    public final ChainType chainType;
+    /**
+     * Holds the entity ids of associated {@link ChainCollisionEntity collision entities}.
+     */
+    private final IntList collisionStorage = new IntArrayList(16);
+    /**
+     * Indicates that no sound should be played when the link is destroyed.
+     */
+    public boolean removeSilently = false;
+    /**
+     * Whether the link exists and is active
+     */
+    private boolean alive = true;
+
+    private ChainLink(@NotNull ChainKnotEntity primary, @NotNull Entity secondary, @NotNull ChainType chainType) {
+        if (primary.equals(secondary))
+            throw new IllegalStateException("Tried to create a link between a knot and itself");
+        this.primary = Objects.requireNonNull(primary);
```

Comment:
Is it not better to log an error and then just not create the chainlink?
Throwing an error is bad as it causes the whole game to stop.

Reply:
When you create a link between a knot and itself it will cause undefined behavior. It think it would cause bigger issues (crash or even infinite loop, I can't tell you) elsewhere which would be hard to track back to this, even with a log message. Code that uses the chain link should be able to assume that `primary == secondary` is **never** true. The param doc of the `create` function explicitly states `A different chain knot or player`.

Reply:
Yes, but can it then not delete the chain completely? 
If you you are sure it sill never happen, then oke we will keep it.